### ISSUE TITLE
Sequence unit formatting

### ISSUE THREAD

**Issue by sffc** [tc39/ecma402#398]
Sequence units are things like "feet and inches", "hours and minutes", etc.

I have a [CLDR proposal](https://docs.google.com/document/d/1BSJFap5SnDX3E4AbG6pdK5S0HFHYQzl1d7OVD3HxP1c/edit#) to extend unit identifier syntax to support sequence units with the `+` separator:

- `foot+inch`
- `meter+centimeter`
- `hour+minute+second`

If we adopted this extension into ECMA-402, then this is how we could format sequence units:

```javascript
const fmt = new Intl.NumberFormat("de-CH", {
    style: "unit",
    unit: "meter+centimeter"
});
console.log(fmt.format(125));
// "1 m 25 cm"
```

@younies @littledan @zbraniecki @longlho

**Comment by sffc** [comment]
Note: currently, you are able to format sequence units by combining the output of Intl.NumberFormat into Intl.ListFormat.

```javascript
const meterFmt = new Intl.NumberFormat("de-CH", {
    style: "unit",
    unit: "meter"
});
const centimeterFmt = new Intl.NumberFormat("de-CH", {
    style: "unit",
    unit: "centimeter"
});
const lf = new Intl.ListFormat("de-CH", {
    type: "unit"
});
console.log(lf.format(meterFmt.format(1), centimeterFmt.format(25)));
```

Currently this does not work in Chrome though?  I get `"1,   und m"`  @FrankYFTang

**Comment by FrankYFTang** [comment]
> Note: currently, you are able to format sequence units by combining the output of Intl.NumberFormat into Intl.ListFormat.
> 
> ```js
> const meterFmt = new Intl.NumberFormat("de-CH", {
>     style: "unit",
>     unit: "meter"
> });
> const centimeterFmt = new Intl.NumberFormat("de-CH", {
>     style: "unit",
>     unit: "centimeter"
> });
> const lf = new Intl.ListFormat("de-CH", {
>     type: "unit"
> });
> console.log(lf.format(meterFmt.format(1), centimeterFmt.format(25)));
> ```
> 
> Currently this does not work in Chrome though? I get `"1, und m"` @FrankYFTang

You call the format of ListFormat incorrectly

```
lf.format([meterFmt.format(1), centimeterFmt.format(25)])
"1 m, 25 cm"
```
You need to wrap that with [ ]

The spec stated:
```
1.4.3Intl.ListFormat.prototype.format ( list )
When the format method is called with an argument list, the following steps are taken:

Let lf be the this value.
If Type(lf) is not Object, throw a TypeError exception.
If lf does not have an [[InitializedListFormat]] internal slot, throw a TypeError exception.
Let stringList be ? StringListFromIterable(list).
Return FormatList(lf, stringList).
```

Notice it only take ONE argument.

**Comment by longlho** [comment]
In general this is great! I'm curious about the decision behind `+` vs a real `List`?

**Comment by sffc** [comment]
Sure, good question.

1. **API clarity:** Let's say you set `unit: ["foot", "inch"]`.  Now you call `.format(71)`.  What happens?  Do you resolve using the smallest unit, like in my `foot+inch` example?  Or do we extend .format to also take a list, like `.format([5, 11])`?
2. **Defer to CLDR:** The CLDR spec would come with a specification for how values in multiple units are resolved when formatting, so ECMA-402 doesn't need to own that decision.
3. **Sequence units are just units:** The model of formatting a "list" of units, as ICU currently does, is confusing and unnecessary.  "foot+inch" is as much of a unit as just "inch"; the only difference is how it's rendered.

**Comment by longlho** [comment]
Are sequence units gonna be subject to a sanction list the same way regular units are right now? Or are we applying the same rule as `unit1-per-unit2` where anything can be combined as long as `unit1` & `unit2` are sanctions?

**Comment by sffc** [comment]
Also a good question.  From a CLDR syntax perspective, arbitrary combinations are OK.  In ICU, the model I was thinking was that the .format() method would throw if the combination was invalid (for example, combining two units that measure different quantities).  In JavaScript, we should probably throw in the constructor, though, not the .format() method.

The requirements for a "valid" sequence unit would be:

1. The units must all measure the same quantity (length, time, acceleration, etc.)
2. From left to right, units must be strictly decreasing in scale ("foot+inch" but not "inch+foot")
3. Anything else?

**Comment by longlho** [comment]
Do we allow gaps? e.g `[hour, second]` can result in `1 hour 200 seconds`?

**Comment by sffc** [comment]
> Do we allow gaps? e.g `[hour, second]` can result in `1 hour 200 seconds`?

Yeah, in this situation gaps are fine.  This is not the same as #394 since durations are intrinsically different than datetimes.

**Comment by zbraniecki** [comment]
for what its worth, it may be too late, but I struggle with the `"meter+centimeter"` notation. It's non-semantic, and requires awkward string parsing/serialization to remodel. I'd much prefer `["meter", "centimeter"]`.

**Comment by longlho** [comment]
Can u also clarify how the algorithms for calculating the units are going to be spec'ed out, or whether they're out of scope? The context for my question is that Apple, on Mac, decides to format bytes as base 10 instead of base 2 (so 1000 instead of 1024) (https://support.apple.com/en-us/HT204612)

**Comment by sffc** [comment]
> it may be too late

It's not too late at all; the CLDR spec has not been approved yet, and even if it does, ECMA-402 could choose to do something different.

> It's non-semantic

Can you elaborate on this more?  I find the string to be semantically accurate for reasons I listed in https://github.com/tc39/ecma402/issues/398#issuecomment-573812164.  I would like to understand better the intuition behind a List being more semantic.

> requires awkward string parsing/serialization to remodel

This is true; my expectation though is that the common case does not involve "building up" these identifiers programatically.  Developers should just pass these identifiers through from some other source, such as an API that returns a unit identifier string given a locale and context.

> Can u also clarify how the algorithms for calculating the units are going to be spec'ed out

I talked with Mark Davis about this.  The solution we had in mind was:

1. The quantity being formatted is expressed in the smallest unit; the spec will include a list of units from largest to smallest in each category.  For example, quantities in "foot+inch" and "inch+foot"* will both be expressed in inches.
2. If there are private-use units in play (not relevant for ECMA-402), the quantity is expressed in the private-use unit that is latest in the alphabet.
3. Conversions between comparable units will be provided in data.
4. When resolving the quantities for formatting, the quantity should be converted to the largest unit, rounded to the nearest integer, and then the remainder converted to the next smaller unit, so on and so forth down to the smallest unit, where the quantity is rounded according to the user-specified rounding rules in the NumberFormat.
5. (maybe?) The implementation may choose to approximate the calculations (e.g., in double space) and still be spec-compliant.

Does that sound robust enough?

\* Certain locales prefer to put smaller units first.  Rather than re-ordering the units in the identifier string, we could specify that the rendered order of "foot+inch" is locale-dependent.  However, CLDR does not have data for reordering sequence units, and it's unclear how it could be spec'd out in a robust way.  The unit identifier is already locale-dependent in the sense that you get "foot" in US and "meter" almost everywhere else, so it seems consistent for "foot+inch" and "inch+foot" to be different identifiers.

**Comment by longlho** [comment]
Thanks for the clarification. In terms of
```
Conversions between comparable units will be provided in data
```
Do you mean user-supplied data, or CLDR-like? I'm trying to determine if this handles the `1000 bytes = 1Kb` vs `1024 bytes = 1Kb` thing.

I like the fact that unit combo declaration being locale-independent!

**Comment by sffc** [comment]
My understanding is that the intent is for UTS 35 to specify the correct definition of the unit when there is ambiguity (such as BTU, Cup, Barel, etc), and CLDR will provide conversion data. I guess if you are consuming the spec but not the data, you can provide your own conversion data, as long as it is consistent with the definitions in UTS 35. I'll need to check with @macchiati though to make sure this is accurate.

UTS 35 will define the SI prefix "kilo" to mean 1E3 of the base unit, which means kilobyte is 1000 bytes. UTS 35 could add the binary prefixes like "kibi" in the future if there is sufficient demand.

**Comment by rxaviers** [comment]
> `"foot+inch"` vs `["foot", "inch"]`...

It's also important to think about its distinction from -per-, i.e.:
- `"foo+inch"` vs `"foo-per-inch"`
- `["foo", "inch"]` vs `"foo-per-inch"`

Both operations: sequence and division are lists of units.

**Comment by sffc** [comment]
The [latest CLDR proposal](https://docs.google.com/document/d/1NbXHVfitfbm0m2c3zftvCYAdLkurwIC5u1W5HHwWYt8/edit) uses "-and-" as an infix, like "-per-", with quantities measured in the largest unit.  So, for example, 4.25 with unit "foot-and-inch" would be rendered as "4 feet and 3 inches".

**Comment by sffc** [comment]
@younies Can you add support for mixed units to the Smart Units proposal?

**Comment by rxaviers** [comment]
>  So, for example, 4.25 with unit "foot-and-inch" would be rendered as "4 feet and 3 inches".

@sffc I am summarizing my understanding to make sure I am in the same page... The proposal is to support sequence unit formatting by using `NumberFormat` and units using `-and-` infix.

```js
new Intl.NumberFormat(undefined, {
  style: "unit",
  unit: "foot-and-inch"
}).format(4.25);
// > "4 ft, 3 in" (en-US)
// > "4 pés e 3 pol." (pt-BR)
```

The equivalent code using existing ListFormat + NumberFormat is:

```js
let value = 4.25;
new Intl.ListFormat(undefined, { style: "long", type: "unit" }).format([
  new Intl.NumberFormat(undefined, {
    style: "unit",
    unit: "foot",
    maximumFractionDigits: 0
  }).format(value),
  new Intl.NumberFormat(undefined, {
    style: "unit",
    unit: "inch"
  }).format(value % 1 * 12)
]);
// > "4 ft, 3 in" (en-US)
// > "4 pés e 3 pol." (pt-BR)
```

Therefore, the requirement to support such proposal is to support Smart Units (i.e., ability to make unit conversions). Without the ability to convert units automatically, `NumberFormat` + `-and-` units isn't possible. Is that correct?

**Comment by sffc** [comment]
@rxaviers 

The code is *almost* right, except that the ListFormat width should match the unit width, and the first value should be floored, not rounded up.

From an i18n functionality perspective, this proposal doesn't carry much weight, that's true. In my mind, it's more about parity with UTS 35. The unit identifier spec has been evolving (starting with your suggestion a couple years back to remove the "length" from "length-meter"), and this is simply the latest change. CLDR contains the conversion data.

Although this proposal is related to smart units, I don't see one as depending on the other.

**Comment by younies** [comment]
> @younies Can you add support for mixed units to the Smart Units proposal?

Sure, I will add that, I am filling an issue right now to remind me.

**Comment by Zearin** [comment]
Stumbled onto this thread after I looked into whether the `Intl` API could handle some (U.S.) measurement formatting I’m currently doing manually.

----

### Aside 🤫 

🙏🏽Thanks everyone here who works on this stuff.  I’m not a natural when it comes to numbers and math, so I _love_ it whenever a new API offers the functionality I need (and I don’t have to second-guess whether I calculated things correctly, or fail in edge cases I don’t know about, etc.)!

----

Not sure if this conversation is still in play, however; the link in [@sffc’s comment](https://github.com/tc39/ecma402/issues/398#issuecomment-588614621) shows a document that is marked as obsolete (with a link directing me to a document that I don’t have permission to read).  

If so, I’m looking forward to it!

If not, I learned the `ListFormat` technique here, which is more than I knew when I started reading.  :)

**Comment by sffc** [comment]
The spec has been finalized in CLDR for some time now:

https://www.unicode.org/reports/tr35/tr35-general.html#62-unit-identifiers

**Comment by B-Scott33** [comment]
> Sequence units are things like "feet and inches", "hours and minutes", etc.
> 
> I have a [CLDR proposal](https://docs.google.com/document/d/1BSJFap5SnDX3E4AbG6pdK5S0HFHYQzl1d7OVD3HxP1c/edit#) to extend unit identifier syntax to support sequence units with the `+` separator:
> 
> - `foot+inch`
> - `meter+centimeter`
> - `hour+minute+second`
> 
> If we adopted this extension into ECMA-402, then this is how we could format sequence units:
> 
> ```javascript
> const fmt = new Intl.NumberFormat("de-CH", {
>     style: "unit",
>     unit: "meter+centimeter"
> });
> console.log(fmt.format(125));
> // "1 m 25 cm"
> ```
> 
> @younies @littledan @zbraniecki @longlho

**Comment by sffc** [comment]
TG2 discussion: https://github.com/tc39/ecma402/blob/main/meetings/notes-2026-03-19.md#sequence-unit-formatting-398

We would like to follow CLDR's precedent as much as possible unless there's a good reason not to. I have homework to investigate some of their decisions such as whether to index the unit by the first, last, largest, or smallest.

**Comment by sffc** [comment]
One key question is whether `.format({ value: 3.65, unit: "foot-and-inch" })` should output "3 feet 7.8 inches" or "0 feet 3.65 inches".

I did some digging and all I could find for CLDR's thinking was this comment from @hugovdm

> Mixed units are basically an alternative to decimal fractions, so it's 3 feet, and 0.65 expressed in inches.

https://github.com/icu-units/icu/issues/88

The primary argument in favor of the other approach is that we can do all of the math in terms of even, integer multiples. For example, from 43.8 inches we can take mod 12 to get 3 feet, with a remainder of 7.8 inches. In the web platform where we only have access to integer and floating point operations, this might be compelling.

**Comment by hugovdm** [comment]
> The primary argument in favor of the other approach is that we can do all of the math in terms of even, integer multiples. For example, from 43.8 inches we can take mod 12 to get 3 feet, with a remainder of 7.8 inches. In the web platform where we only have access to integer and floating point operations, this might be compelling.

That's a compelling argument, I regret not raising it when working on ICU4C. ☺️

How much does a behaviour mismatch matter?

**Comment by eemeli** [comment]
One prior art that we should consider is Intl.DurationFormat, and its [`duration`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/DurationFormat/format#duration) objects. By this I mean representing compound unit values as objects, as in `{ foot: 3, inch: 7.8 }`.

**Comment by sffc** [comment]
A "flat" approach has a smaller surface:

```javascript
numberFormat.format({
    value: 43.8,
    unit: "foot-and-inch"
})
```

The only real design question is whether `value` is feet or inches.

If we go with a "structured" approach, we have a bigger design space. Here are a bunch of different ways we could write the call site:

```javascript
// A
numberFormat.format({
    value: { foot: 3, inch: 7.8 },
    unit: "foot-and-inch"
})

// B
numberFormat.format({
    value: { foot: 3, inch: 7.8 },
    unit: ["foot", "inch"]
})

// C
numberFormat.format({
    value: [3, 7.8],
    unit: ["foot", "inch"]
})

// D
numberFormat.format({
    foot: 3,
    inch: 7.8,
})

// E
numberFormat.format({
    value: { foot: 3, inch: 7.8 },
    unit: "infer"
})
```

**Comment by gibson042** [comment]
I find it extremely unfortunate that Unicode settled on "$bigger-and-$smaller", because `numberFormat.format({ value, unit: "foot-and-inch" })` really looks like `value` is primarily in feet (e.g., that a value of 3.65 is between 3 feet and 4 feet). But that being the case, I don't think we can diverge. However, we probably **can** still hope for mitigation via conversion (_e.g. `new Intl.NumberFormat("en", { style: "unit", unit: "foot-and-inch" }).format({ value: 43.8, unit: "inch" })` producing "3 feet 7.8 inches")_.

**Comment by sffc** [comment]
> I find it extremely unfortunate that Unicode settled on "$bigger-and-$smaller", because `numberFormat.format({ value, unit: "foot-and-inch" })` really looks like `value` is primarily in feet (e.g., that a value of 3.65 is between 3 feet and 4 feet).

Can you explain your intuition here? What makes `foot-and-inch` "look like" it is measured primarily in feet?

I don't personally have an intuition on which one is more correct. If you have such an intuition, then perhaps it's a question we should ask in a survey.

**Comment by eemeli** [comment]
I tried taking another look at how ICU handles mixed units, and tbh I could not figure it out. As in, I can't tell how you might construct a MeasureUnit that matches e.g. `foot-and-inch` for use with NumberFormatter's `.unit()`. A code example for how one uses ICU to format a `foot-and-inch` value would be quite useful. As would examples for how it's done with other libraries.

If it turns out that mixed unit formatting is more novel than we've been presuming, maybe it's appropriate to take heed of how in CLDR's [units.xml](https://github.com/unicode-org/cldr/blob/main/common/supplemental/units.xml), unit conversion is not in fact specified for any `-and-`, but only for single units, as in
```xml
<convertUnit source='inch' baseUnit='meter' factor='ft_to_m/12' systems="ussystem uksystem"/>
<convertUnit source='foot' baseUnit='meter' factor='ft_to_m' systems="ussystem uksystem"/>
```

We may want to consider initially leaving out locale-sensitive unit conversion from Amount unless there's decent prior art for it. And if we do that, that proposal is no longer blocked on progress here.

**Comment by sffc** [comment]
I would not describe mixed unit formatting as any more "novel" than most of the rest of the localized unit formatting and conversion machinery.

It is reachable in the unitPreferences data:

```xml
        <unitPreferences category="length" usage="person-height">
            <unitPreference regions="001">centimeter</unitPreference>
            <unitPreference regions="CA GB IN US" geq="3.0">foot-and-inch</unitPreference>
            <unitPreference regions="CA GB IN US">inch</unitPreference>
            <unitPreference regions="AT BE DZ EG ES FR HK ID IL IT JO MY SA SE TR VN">meter-and-centimeter</unitPreference>
        </unitPreferences>
```

It is not in the convertUnit data because that is defined for single units; mixed units are a sequence of single units.

---

Although I disagree with the characterization of "novelty" or "lack of prior art" as a reason to leave out locale-aware unit conversion, I am open to smaller proposals, and I think adding locale-aware unit conversion on top of an Amount that supports non-locale-aware unit conversion would be feasible, with caveats:

1. Locale-aware unit conversion is the ultimate goal and the primary motivator for this line of work. I am of course personally still happy with Amount, but removing this piece of functionality might bring us back to a world where some see Amount as less clearly motivated as a standalone proposal.
2. I wouldn't want the locale-aware piece to be blocked later.

**Comment by gibson042** [comment]
> Can you explain your intuition here? What makes `foot-and-inch` "look like" it is measured primarily in feet?

Basically, because it reads like "and inch" refines "foot". Imagine placing an order at a restaurant, e.g. "patty melt... and a cup of soup". I can probably elaborate further if you like, but this intuition is very strong, and probably for more than one reason.

**Comment by eemeli** [comment]
> I would not describe mixed unit formatting as any more "novel" than most of the rest of the localized unit formatting and conversion machinery.

If that's the case, how can I construct an ICU MeasureUnit representing `foot-and-inch`?

**Comment by Zearin** [comment]
> Basically, because it reads like "and inch" refines "foot". 

Hmm…

Well, decimal system gets around this easily by just moving a decimal point or switching units. 

We barbarians in the United States grow up used to no such convenience, which is why a feet and inches combo is so necessary. 

> Imagine placing an order at a restaurant, e.g. "patty melt... and a cup of soup". I can probably elaborate further if you like, but this intuition is very strong, and probably for more than one reason.

Actually that example confuses me. Could you go ahead and elaborate?

**Comment by anba** [comment]
> If that's the case, how can I construct an ICU MeasureUnit representing `foot-and-inch`?

@eemeli  This example ICU program will print `result: 3 ft, 7.8 in`. Is that what you were looking for? 

<details>

```cpp
#include <string_view>

#include "unicode/unumberformatter.h"
#include "unicode/utypes.h"
#include "unicode/ustdio.h"

#define RETURN_ON_FAILURE(status) if (U_FAILURE(status)) { printf("err: %s\n", u_errorName(status)); return 1; }

int main() {
  const char* locale = "en";
  double value = 3.65;
  std::u16string_view skeleton = u"unit/foot-and-inch";

  UErrorCode status = U_ZERO_ERROR;

  auto* nf = unumf_openForSkeletonAndLocale(skeleton.data(), skeleton.length(), locale, &status);
  RETURN_ON_FAILURE(status);

  auto* result = unumf_openResult(&status);
  RETURN_ON_FAILURE(status);

  unumf_formatDouble(nf, value, result, &status);
  RETURN_ON_FAILURE(status);

  auto* formattedValue = unumf_resultAsValue(result, &status);
  RETURN_ON_FAILURE(status);

  int32_t length;
  const char16_t* str = ufmtval_getString(formattedValue, &length, &status);
  RETURN_ON_FAILURE(status);

  u_printf("result: %S\n", str);
}
```

</details>

**Comment by gibson042** [comment]
> > Imagine placing an order at a restaurant, e.g. "patty melt... and a cup of soup". I can probably elaborate further if you like, but this intuition is very strong, and probably for more than one reason.
> 
> Actually that example confuses me. Could you go ahead and elaborate?

"and" is an additive operation; it tacks on additional information to a grammatical object that is presumptively valid on its own. In the above example, the order could have ended with "patty melt", but after a pause was extended by "and" in a way that did not affect the already-established request for a patty melt. Analogously, 3.65 is valid with unit "foot" and its interpretation should not be affected by appending "-and-inch" to that unit, and likewise (to demonstrate that this is not limited to United States customary units) with unit "degree" vs. "degree-and-arc-minute"/"degree-and-arc-minute-and-arc-second", "meter" vs. "meter-and-centimeter", and "minute" vs. "minute-and-second".

**Comment by macchiati** [comment]
For a compound unit like foot-and-inch, the value is a _pair_ of numbers (or triple, etc). So
```
(17, 2.5) foot-and-inch
=> 
5.2451 meters
=>
(5, 2.2083) yard-and-foot
```
Internally to support formatting and user preferences across languages, ICU support converting from simple units (like meters) to either simple units (like fathoms) or compound units like foot-and-inch. A proposal for implementing it was approved for implementation, but I don't know whether it was ever finished. @markusicu would know more.

Once you have simple unit conversion, handling compound unit is not particularly difficult. The only tricky issue is handling precision.

**Comment by eemeli** [comment]
> > If that's the case, how can I construct an ICU MeasureUnit representing `foot-and-inch`?
> 
> [@eemeli](https://github.com/eemeli) This example ICU program will print `result: 3 ft, 7.8 in`. Is that what you were looking for?

Thanks, that's just the example I was looking for. Is the skeleton string the only way to use it? And do you know if there's a reason why MeasureUnit seems to support all the other unit types, but not mixed units?

**Comment by sffc** [comment]
On the structured approach: we discussed this in the Numerics Champions meeting. Some thoughts on the shape:

- `Amount`'s `value` field is an object with fields corresponding to the single units in the unit identifier.
- If any of the units of the unit identifier are missing from the value object, it is an error.
- If there are more units in the value object than in the unit identifier, they are ignored.
- Sequence units _require_ an object of this shape. If a primitive is given, it is an error.
- When a sequence unit is the input to `convertTo`, we convert and add the fields.
- Not yet clear on how `Amount.prototype.toString` behaves on these.

**Comment by macchiati** [comment]
Sounds good.

**Comment by sffc** [comment]
> * Not yet clear on how `Amount.prototype.toString` behaves on these.

We discussed this more in the numerics call today. Here are some solutions with pros and cons:

1. Serialize into a scalar, and deserialize into structured sequence units: "3ft 6in" <=> `{ value: { foot: 3, inch: 6 }, unit: "foot-and-inch" }` <=> "3.5[foot-and-inch]"
    - Pro: Compatible with existing syntax
    - Con: Doesn't support userland mixed units where we don't have conversion factors
    - Observation: Doesn't support "3ft 18in"
2. Serialize as `3,6[foot-and-inch]` or similar (keep general annotation structure)
    - Con: `parseFloat` doesn't see the inch value
3. Serialize as `[3 foot, 6 inch]` or similar (departure from annotation structure)
    - Con: unitless Amounts end up as `[3]` or perhaps `[3 -]` which looks like an array
    - Con: annotations are good and this sets us back in considering how to serialize them
4. Serialize as `=3,6[foot-and-inch]` or similar (keep annotations, but prevent parseFloat)
    - Might be a good compromise

This discussion also raises questions about whether we should include serialization in the proposal. With a `.value` field that supports strings, we don't _need_ toString for interop purposes _as much_. However, even if we leave it out of scope, we should have a general idea of whether the problem is solvable.

**Comment by eemeli** [comment]
@sffc Let's continue in https://github.com/tc39/proposal-amount/issues/97.

**Comment by sffc** [comment]
Sequence unit formatting is composable with NumberFormat and ListFormat, similar to DurationFormat. But unlike DurationFormat, sequence unit formatting in CLDR does not currently support any narrow/numeric styles. 

| Unit | Example | Supported by CLDR? |
|---|---|---|
| hour-and-minute-and-second | 05:30:15 | Using a dedicated pattern |
| foot-and-inch | 5' 11'' | Using ListFormat |
| degree-and-arcminute-and-arcsecond | 30° 45' 22'' | Using ListFormat |

Thought: In principle, Amount can run the ListFormat in its toLocaleString. We don't _need_ sequence unit formatting in Intl.NF so long as CLDR doesn't support numeric formats for inches and arcminutes.

**Comment by eemeli** [comment]
> We don't _need_ sequence unit formatting in Intl.NF so long as CLDR doesn't support numeric formats for inches and arcminutes.

I don't understand how a CLDR capability being introduced would expand what is needed in Intl.NumberFormat.

I also don't see how the listed units are not supported by CLDR, esp. given how the narrow form of the degree-and-arc-minute-and-arc-second representation is used in the LDML as an [example](https://unicode.org/reports/tr35/tr35-info.html#mixed-units:~:text=1%C2%B0%2031%E2%80%B2%2031%2E44) for mixed units, and the narrow form of foot-and-inch as a [example](https://unicode.org/reports/tr35/tr35-general.html#ListPatterns:~:text=3%E2%80%B2%207) for list patterns.

Is it perhaps an ICU lack of capability that you're observing?

**Comment by sffc** [comment]
I see a small number of locales that use the numeric formats as their unit pattern. Here's one:

https://github.com/unicode-org/cldr/blob/390a8bdb504f771bda01f2f0a1d412640afd75b4/common/main/ig.xml#L8418

I don't see this used in the context of alternate long, short, and narrow forms, but it would be easy for locales to do this. Once they do, the pieces are glued back together using a ListFormat.

The thing CLDR doesn't have, as far as I see, are patterns for mixed units that bypass ListFormat entirely. Something that would look like:

```
<unit-pattern unit="foot-and-inch">{0}′ {1}″</unit-pattern>
```

Since mixed units always go through ListFormat, then NumberFormat supporting them natively is not a value-add on the web platform.

**Comment by sffc** [comment]
TG2 discussion on this topic: https://github.com/tc39/ecma402/blob/main/meetings/notes-2026-04-16.md#sequence-units-follow-up

There was a desire for an Amount to be able to be passed to an Intl formatter and have it "just work", rather than relying on `toLocaleString` to implement the logic. Therefore, it seems TG2 prefers to move forward with this proposal even though it is implementable on top of existing 402 code (ListFormat).

**Comment by sffc** [comment]
Proposal repo: https://github.com/tc39-transfer/proposal-intl-sequence-units

I'll spawn issues for the individual discussions. We should move further discussion to the proposal repo and keep this issue only for updates and proposal-wide decisions.
